### DISCUSSION TITLE
Heatmap Layer incorrect color mapping

### DISCUSSION THREAD

**Discussion by amanfucci** [visgl/deck.gl#5778] [Q&A] [answered]
**Description**
The implemented heatmap layer incorrectly maps any weight to the highest level colour.
_d.fire_index_ stores values from 0 to 5, hence the weight of each point has a value from 1 to 6.
By setting the colorDomain to [0.02, 5] only the points with the weight of 6 should get mapped to the highest colour (purple)
```
my_map.setProps({
    layers: [
        new deck.HeatmapLayer({
            id: 'heat-map',
            data: data,
            opacity: 0.3,
            intensity: 1,
            getPosition: d => [parseFloat(d.lat), parseFloat(d.lng)],
            getWeight: d => d.fire_index+1,
            colorDomain: [0.02, 5],
            radiusPixels: 40,
            colorRange: [[0, 172, 105], [244, 161, 0], [247, 100, 0], [232, 21, 0], [227, 0, 89], [105, 0, 99]],
            aggregation: 'MEAN'
        })]
})
```
![Capture](https://user-images.githubusercontent.com/36966553/118252355-f56d0480-b4a8-11eb-8ccc-4d38b55f1957.PNG)
**Environment**
- Framework: DeckGl 8.4.16 Standalone (Scripting API)
- Browser: Opera 76.0.4017.107, Microsoft Edge 90.0.818.56
- OS: Windows 10

**Comment by sclebal** [comment]
Hi @almanfu,
At first glance, I don't see anything that can be causing your error.
Can you provide a reproduction example for looking into it deeper?

**Reply by amanfucci** [reply]
Hi @Clebal , thanks for your response.
The codepen below reproduces the issue.
Looking at other discussions (https://github.com/visgl/deck.gl/issues/4005) It seems that this issue is rather the way deck.gl heatmaps work (by handling only discrete values such that setting a point weight to more than 1 is really just adding more points to the same location). Thus they cannot be used for representing actual temperature.
[https://codepen.io/manfu/pen/abJmNaX](https://codepen.io/manfu/pen/abJmNaX)

**Reply by sclebal** [reply]
Hi @almanfu, I don't know if I'm far from the solution, but have you tried to put every posible weight value in `colorDomain`? 
Using: `colorDomain: [0, 1, 2, 3, 4, 5],` I see this:

![image](https://user-images.githubusercontent.com/9151432/118674568-2aa38a80-b7fa-11eb-8d99-e41a70da0f9c.png)

Is it what you want to achieve? Maybe I'm not fully understanding you.

**Reply by amanfucci** [reply]
Hi @Clebal, It seems that by setting the colorDomain to an array with more than two elements, deck.gl overrides the colorDomain with its default settings ([deck.gl/docs/api-reference/aggregation-layers/heatmap-layer](https://deck.gl/docs/api-reference/aggregation-layers/heatmap-layer)).

Still, the problem lies in the way heatmaps are created.
For instance, by using the default colorDomain a point seems to change its color depending on the zoom level, which makes these heatmaps not suitable for this use (which is really just displaying temperatures):

- Zoomed Out

<img src="https://user-images.githubusercontent.com/36966553/118679719-693b4400-b7fe-11eb-86c5-fa83c1418266.PNG" width="700"/>

- Zoomed In
<img src="https://user-images.githubusercontent.com/36966553/118679718-68a2ad80-b7fe-11eb-8bff-48994cf398dd.PNG" width="700"/>

**Reply by Pessimistress** [reply]
The layer can definitely be used to display temperature. If you remove `colorDomain` it works as expected. I will reopen your issue.

**Comment by Pessimistress** [comment, marked as answer]
@almanfu I've opened PR https://github.com/visgl/deck.gl/pull/5802 to address the issue and attempt to clarify the usage of `colorDomain` in the documentation. Would appreciate your feedback!

**Reply by amanfucci** [reply]
Hi @Pessimistress, the new documentation extensively clarifies the usage of ColorDomain.
Can't wait to test the new commit!
